{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recruitment values from traditional viruscope and batch viruscope aren't matching... want to figure out why in this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exesag = \"AG-903-I06\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import os.path as op\n",
    "from recruitment_for_vs import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "odpovtsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_vs_POV.tsv.gz\".format(exesag)\n",
    "odlineptsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_vs_LineP-all.tsv.gz\".format(exesag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pov = import_diamond_tsv(odpovtsv, best_hit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def construct_recruit_tbl(vir_tsv, bac_tsv, read_count_dict, contig_file):\n",
    "    '''\n",
    "    Args:\n",
    "        vir_tsv: diamond recruitment converted to tsv for vir metagenome\n",
    "        bac_tsv: diamond recruitment converted to tsv for bac metagenome\n",
    "        read_count_dict: dict of mg read counts with two keys -- 'vir_reads' and 'bac_reads'\n",
    "        contig_file: path to a file with sag contigs in it; either in fasta or gff format\n",
    "    Returns:\n",
    "        pandas dataframe with mg fraction calculated\n",
    "    '''\n",
    "    cnames = \"qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore\".split()\n",
    "    bac_df = import_diamond_tsv(bac_tsv)\n",
    "    vir_df = import_diamond_tsv(vir_tsv)\n",
    "\n",
    "    bac_sum = pd.Series(bac_df.groupby('sseqid')['qseqid'].count(), name='hit_mg-bac')\n",
    "    vir_sum = pd.Series(vir_df.groupby('sseqid')['qseqid'].count(), name='hit_mg-vir')\n",
    "\n",
    "    orfhits = pd.concat([bac_sum, vir_sum], axis=1).reset_index().rename(columns={'index':'orf'})\n",
    "    orfhits = map_orfs_to_contigs(orfhits, contig_file)\n",
    "    \n",
    "    chits = pd.concat([summarize_by_contig(orfhits, 'hit_mg-bac'), summarize_by_contig(orfhits, 'hit_mg-vir')], axis=1)\n",
    "    chits['reads_mg-vir'] = float(read_count_dict['vir_reads'])\n",
    "    chits['reads_mg-bac'] = float(read_count_dict['bac_reads'])\n",
    "    \n",
    "    clens = contig_lengths(contig_file)\n",
    "    \n",
    "    out_tbl = compute_fr(chits.reset_index(), clens, mult=1e6)\n",
    "    \n",
    "    return out_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# constructing recruit table from vs test run :\n",
    "\n",
    "vtsv = \"/mnt/scgc/simon/simonsproject/jb_vs_test/AG-903/AG-903-I06/diamond/POV.tsv.gz\"\n",
    "btsv =  \"/mnt/scgc/simon/simonsproject/jb_vs_test/AG-903/AG-903-I06/diamond/LineP-all.tsv.gz\"\n",
    "read_count_dict = {'vir_reads':5922080, 'bac_reads':8279226}\n",
    "contig_file = \"/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/AG-903/AG-903-I06_contigs.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "assert op.exists(contig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n"
     ]
    }
   ],
   "source": [
    "df1 = construct_recruit_tbl(vtsv, btsv, read_count_dict, contig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>hit_mg-bac</th>\n",
       "      <th>hit_mg-vir</th>\n",
       "      <th>reads_mg-vir</th>\n",
       "      <th>reads_mg-bac</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-903-I06_NODE_1</td>\n",
       "      <td>555.0</td>\n",
       "      <td>1551.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>35014.0</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>0.007480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-903-I06_NODE_10</td>\n",
       "      <td>21.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>12198.0</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.002049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-903-I06_NODE_11</td>\n",
       "      <td>393.0</td>\n",
       "      <td>189.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>10789.0</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>0.002958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-903-I06_NODE_12</td>\n",
       "      <td>285.0</td>\n",
       "      <td>3194.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>10602.0</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.050871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-903-I06_NODE_13</td>\n",
       "      <td>57.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>10047.0</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG-903-I06_NODE_14</td>\n",
       "      <td>479.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>9782.0</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.000069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG-903-I06_NODE_15</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>7199.0</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG-903-I06_NODE_16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>5516.0</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG-903-I06_NODE_17</td>\n",
       "      <td>19.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>5199.0</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG-903-I06_NODE_18</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>3460.0</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AG-903-I06_NODE_2</td>\n",
       "      <td>3270.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>30265.0</td>\n",
       "      <td>0.013050</td>\n",
       "      <td>0.001685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AG-903-I06_NODE_20</td>\n",
       "      <td>46.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>2400.0</td>\n",
       "      <td>0.002315</td>\n",
       "      <td>0.010765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AG-903-I06_NODE_3</td>\n",
       "      <td>4115.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>27205.0</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>0.001012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AG-903-I06_NODE_4</td>\n",
       "      <td>62.0</td>\n",
       "      <td>732.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>24288.0</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.005089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AG-903-I06_NODE_5</td>\n",
       "      <td>2138.0</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>21277.0</td>\n",
       "      <td>0.012137</td>\n",
       "      <td>0.009270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AG-903-I06_NODE_6</td>\n",
       "      <td>206.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>16072.0</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AG-903-I06_NODE_7</td>\n",
       "      <td>414.0</td>\n",
       "      <td>2161.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>14765.0</td>\n",
       "      <td>0.003387</td>\n",
       "      <td>0.024714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AG-903-I06_NODE_8</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>13309.0</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AG-903-I06_NODE_9</td>\n",
       "      <td>791.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>12540.0</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.000525</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig  hit_mg-bac  hit_mg-vir  reads_mg-vir  reads_mg-bac  \\\n",
       "0    AG-903-I06_NODE_1       555.0      1551.0     5922080.0     8279226.0   \n",
       "1   AG-903-I06_NODE_10        21.0       148.0     5922080.0     8279226.0   \n",
       "2   AG-903-I06_NODE_11       393.0       189.0     5922080.0     8279226.0   \n",
       "3   AG-903-I06_NODE_12       285.0      3194.0     5922080.0     8279226.0   \n",
       "4   AG-903-I06_NODE_13        57.0         NaN     5922080.0     8279226.0   \n",
       "5   AG-903-I06_NODE_14       479.0         4.0     5922080.0     8279226.0   \n",
       "6   AG-903-I06_NODE_15        43.0         1.0     5922080.0     8279226.0   \n",
       "7   AG-903-I06_NODE_16         1.0         NaN     5922080.0     8279226.0   \n",
       "8   AG-903-I06_NODE_17        19.0         2.0     5922080.0     8279226.0   \n",
       "9   AG-903-I06_NODE_18         2.0         NaN     5922080.0     8279226.0   \n",
       "10   AG-903-I06_NODE_2      3270.0       302.0     5922080.0     8279226.0   \n",
       "11  AG-903-I06_NODE_20        46.0       153.0     5922080.0     8279226.0   \n",
       "12   AG-903-I06_NODE_3      4115.0       163.0     5922080.0     8279226.0   \n",
       "13   AG-903-I06_NODE_4        62.0       732.0     5922080.0     8279226.0   \n",
       "14   AG-903-I06_NODE_5      2138.0      1168.0     5922080.0     8279226.0   \n",
       "15   AG-903-I06_NODE_6       206.0        12.0     5922080.0     8279226.0   \n",
       "16   AG-903-I06_NODE_7       414.0      2161.0     5922080.0     8279226.0   \n",
       "17   AG-903-I06_NODE_8         3.0         NaN     5922080.0     8279226.0   \n",
       "18   AG-903-I06_NODE_9       791.0        39.0     5922080.0     8279226.0   \n",
       "\n",
       "    contig_length  fr_mg-bac  fr_mg-vir  \n",
       "0         35014.0   0.001915   0.007480  \n",
       "1         12198.0   0.000208   0.002049  \n",
       "2         10789.0   0.004400   0.002958  \n",
       "3         10602.0   0.003247   0.050871  \n",
       "4         10047.0   0.000685        NaN  \n",
       "5          9782.0   0.005915   0.000069  \n",
       "6          7199.0   0.000721   0.000023  \n",
       "7          5516.0   0.000022        NaN  \n",
       "8          5199.0   0.000441   0.000065  \n",
       "9          3460.0   0.000070        NaN  \n",
       "10        30265.0   0.013050   0.001685  \n",
       "11         2400.0   0.002315   0.010765  \n",
       "12        27205.0   0.018270   0.001012  \n",
       "13        24288.0   0.000308   0.005089  \n",
       "14        21277.0   0.012137   0.009270  \n",
       "15        16072.0   0.001548   0.000126  \n",
       "16        14765.0   0.003387   0.024714  \n",
       "17        13309.0   0.000027        NaN  \n",
       "18        12540.0   0.007619   0.000525  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "realvs = pd.read_csv(\"/mnt/scgc/simon/simonsproject/jb_vs_test/AG-903/AG-903-I06/summary/AG-903-I06_contigs-summary.csv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'contig_length', 'gene_count', 'viral_phage_gene_count',\n",
       "       'viral_phage_gene_fraction', 'viral2_phage_gene_count',\n",
       "       'viral2_phage_gene_fraction', 'Similarity_1.LineP.all.fr',\n",
       "       'Similarity_1.POV.fr', 'ratio_virus_bacteria', 'virus_class',\n",
       "       'virus_prob'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "realvs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/scgc_nfs/opt/common/anaconda3/4.0.0/lib/python3.5/site-packages/pandas/core/frame.py:2748: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  **kwargs)\n"
     ]
    }
   ],
   "source": [
    "rvs = realvs[['Unnamed: 0','Similarity_1.LineP.all.fr','Similarity_1.POV.fr', 'ratio_virus_bacteria']]\n",
    "rvs.rename(columns = {'Unnamed: 0':'contig'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>Similarity_1.LineP.all.fr</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>Similarity_1.POV.fr</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-903-I06_NODE_1</td>\n",
       "      <td>0.001918</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>0.007610</td>\n",
       "      <td>0.007480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-903-I06_NODE_2</td>\n",
       "      <td>0.013050</td>\n",
       "      <td>0.013050</td>\n",
       "      <td>0.001685</td>\n",
       "      <td>0.001685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-903-I06_NODE_3</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>0.001012</td>\n",
       "      <td>0.001012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-903-I06_NODE_4</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.005089</td>\n",
       "      <td>0.005089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-903-I06_NODE_5</td>\n",
       "      <td>0.012137</td>\n",
       "      <td>0.012137</td>\n",
       "      <td>0.009270</td>\n",
       "      <td>0.009270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG-903-I06_NODE_6</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG-903-I06_NODE_7</td>\n",
       "      <td>0.003379</td>\n",
       "      <td>0.003387</td>\n",
       "      <td>0.024417</td>\n",
       "      <td>0.024714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG-903-I06_NODE_8</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG-903-I06_NODE_9</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.000525</td>\n",
       "      <td>0.000525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG-903-I06_NODE_10</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.002049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AG-903-I06_NODE_11</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>0.002958</td>\n",
       "      <td>0.002958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AG-903-I06_NODE_12</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.050855</td>\n",
       "      <td>0.050871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AG-903-I06_NODE_13</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AG-903-I06_NODE_14</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AG-903-I06_NODE_15</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AG-903-I06_NODE_16</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AG-903-I06_NODE_17</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AG-903-I06_NODE_18</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AG-903-I06_NODE_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AG-903-I06_NODE_20</td>\n",
       "      <td>0.002315</td>\n",
       "      <td>0.002315</td>\n",
       "      <td>0.010765</td>\n",
       "      <td>0.010765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AG-903-I06_NODE_21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig  Similarity_1.LineP.all.fr  fr_mg-bac  \\\n",
       "0    AG-903-I06_NODE_1                   0.001918   0.001915   \n",
       "1    AG-903-I06_NODE_2                   0.013050   0.013050   \n",
       "2    AG-903-I06_NODE_3                   0.018270   0.018270   \n",
       "3    AG-903-I06_NODE_4                   0.000308   0.000308   \n",
       "4    AG-903-I06_NODE_5                   0.012137   0.012137   \n",
       "5    AG-903-I06_NODE_6                   0.001548   0.001548   \n",
       "6    AG-903-I06_NODE_7                   0.003379   0.003387   \n",
       "7    AG-903-I06_NODE_8                   0.000027   0.000027   \n",
       "8    AG-903-I06_NODE_9                   0.007619   0.007619   \n",
       "9   AG-903-I06_NODE_10                   0.000208   0.000208   \n",
       "10  AG-903-I06_NODE_11                   0.004400   0.004400   \n",
       "11  AG-903-I06_NODE_12                   0.003247   0.003247   \n",
       "12  AG-903-I06_NODE_13                   0.000685   0.000685   \n",
       "13  AG-903-I06_NODE_14                   0.005915   0.005915   \n",
       "14  AG-903-I06_NODE_15                   0.000721   0.000721   \n",
       "15  AG-903-I06_NODE_16                   0.000022   0.000022   \n",
       "16  AG-903-I06_NODE_17                   0.000441   0.000441   \n",
       "17  AG-903-I06_NODE_18                   0.000070   0.000070   \n",
       "18  AG-903-I06_NODE_19                        NaN        NaN   \n",
       "19  AG-903-I06_NODE_20                   0.002315   0.002315   \n",
       "20  AG-903-I06_NODE_21                        NaN        NaN   \n",
       "\n",
       "    Similarity_1.POV.fr  fr_mg-vir  \n",
       "0              0.007610   0.007480  \n",
       "1              0.001685   0.001685  \n",
       "2              0.001012   0.001012  \n",
       "3              0.005089   0.005089  \n",
       "4              0.009270   0.009270  \n",
       "5              0.000126   0.000126  \n",
       "6              0.024417   0.024714  \n",
       "7                   NaN        NaN  \n",
       "8              0.000525   0.000525  \n",
       "9              0.002049   0.002049  \n",
       "10             0.002958   0.002958  \n",
       "11             0.050855   0.050871  \n",
       "12                  NaN        NaN  \n",
       "13             0.000069   0.000069  \n",
       "14             0.000023   0.000023  \n",
       "15                  NaN        NaN  \n",
       "16             0.000065   0.000065  \n",
       "17                  NaN        NaN  \n",
       "18                  NaN        NaN  \n",
       "19             0.010765   0.010765  \n",
       "20                  NaN        NaN  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rvs.merge(df1, on='contig', how='outer')[['contig', 'Similarity_1.LineP.all.fr','fr_mg-bac','Similarity_1.POV.fr','fr_mg-vir']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When using the same input, python and R functions generate the same recruit table.  Why, then, are the recruit tables different between the batch version and the normal version?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "batch_tbl = pd.read_csv(\"/mnt/scgc/simon/simonsproject/jb_vs_test/AG-903-I06_new_script.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>Similarity_1.LineP.all.fr</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>Similarity_1.POV.fr</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-903-I06_NODE_1</td>\n",
       "      <td>0.001918</td>\n",
       "      <td>0.000604</td>\n",
       "      <td>0.007610</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-903-I06_NODE_2</td>\n",
       "      <td>0.013050</td>\n",
       "      <td>0.011669</td>\n",
       "      <td>0.001685</td>\n",
       "      <td>0.002154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-903-I06_NODE_3</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>0.017506</td>\n",
       "      <td>0.001012</td>\n",
       "      <td>0.000689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-903-I06_NODE_4</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.005089</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-903-I06_NODE_5</td>\n",
       "      <td>0.012137</td>\n",
       "      <td>0.008243</td>\n",
       "      <td>0.009270</td>\n",
       "      <td>0.008587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG-903-I06_NODE_6</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.002450</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG-903-I06_NODE_7</td>\n",
       "      <td>0.003379</td>\n",
       "      <td>0.004998</td>\n",
       "      <td>0.024417</td>\n",
       "      <td>0.027699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG-903-I06_NODE_8</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG-903-I06_NODE_9</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.007513</td>\n",
       "      <td>0.000525</td>\n",
       "      <td>0.000592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG-903-I06_NODE_10</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.000055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AG-903-I06_NODE_11</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>0.004624</td>\n",
       "      <td>0.002958</td>\n",
       "      <td>0.003287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AG-903-I06_NODE_12</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.002279</td>\n",
       "      <td>0.050855</td>\n",
       "      <td>0.024193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AG-903-I06_NODE_13</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AG-903-I06_NODE_14</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.002210</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AG-903-I06_NODE_15</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AG-903-I06_NODE_16</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AG-903-I06_NODE_17</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000465</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AG-903-I06_NODE_18</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AG-903-I06_NODE_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AG-903-I06_NODE_20</td>\n",
       "      <td>0.002315</td>\n",
       "      <td>0.002818</td>\n",
       "      <td>0.010765</td>\n",
       "      <td>0.009991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AG-903-I06_NODE_21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig  Similarity_1.LineP.all.fr  fr_mg-bac  \\\n",
       "0    AG-903-I06_NODE_1                   0.001918   0.000604   \n",
       "1    AG-903-I06_NODE_2                   0.013050   0.011669   \n",
       "2    AG-903-I06_NODE_3                   0.018270   0.017506   \n",
       "3    AG-903-I06_NODE_4                   0.000308   0.000035   \n",
       "4    AG-903-I06_NODE_5                   0.012137   0.008243   \n",
       "5    AG-903-I06_NODE_6                   0.001548   0.002450   \n",
       "6    AG-903-I06_NODE_7                   0.003379   0.004998   \n",
       "7    AG-903-I06_NODE_8                   0.000027   0.000118   \n",
       "8    AG-903-I06_NODE_9                   0.007619   0.007513   \n",
       "9   AG-903-I06_NODE_10                   0.000208   0.000119   \n",
       "10  AG-903-I06_NODE_11                   0.004400   0.004624   \n",
       "11  AG-903-I06_NODE_12                   0.003247   0.002279   \n",
       "12  AG-903-I06_NODE_13                   0.000685   0.000385   \n",
       "13  AG-903-I06_NODE_14                   0.005915   0.002210   \n",
       "14  AG-903-I06_NODE_15                   0.000721   0.000050   \n",
       "15  AG-903-I06_NODE_16                   0.000022   0.000022   \n",
       "16  AG-903-I06_NODE_17                   0.000441   0.000465   \n",
       "17  AG-903-I06_NODE_18                   0.000070   0.000035   \n",
       "18  AG-903-I06_NODE_19                        NaN   0.000000   \n",
       "19  AG-903-I06_NODE_20                   0.002315   0.002818   \n",
       "20  AG-903-I06_NODE_21                        NaN   0.000000   \n",
       "\n",
       "    Similarity_1.POV.fr  fr_mg-vir  \n",
       "0              0.007610   0.000757  \n",
       "1              0.001685   0.002154  \n",
       "2              0.001012   0.000689  \n",
       "3              0.005089   0.000021  \n",
       "4              0.009270   0.008587  \n",
       "5              0.000126   0.000231  \n",
       "6              0.024417   0.027699  \n",
       "7                   NaN   0.000013  \n",
       "8              0.000525   0.000592  \n",
       "9              0.002049   0.000055  \n",
       "10             0.002958   0.003287  \n",
       "11             0.050855   0.024193  \n",
       "12                  NaN   0.000017  \n",
       "13             0.000069   0.000293  \n",
       "14             0.000023   0.000141  \n",
       "15                  NaN   0.000000  \n",
       "16             0.000065   0.000032  \n",
       "17                  NaN   0.000000  \n",
       "18                  NaN   0.000000  \n",
       "19             0.010765   0.009991  \n",
       "20                  NaN   0.000000  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rvs.merge(batch_tbl, on='contig', how='outer')[['contig', 'Similarity_1.LineP.all.fr','fr_mg-bac','Similarity_1.POV.fr','fr_mg-vir']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOT THE SAME between the two.  Maybe it's an issue with the input files used.  I am using the prokka generated ORFS instead of creating them myself via prodigal.  I'll run a small test on the vs-generated prodigal orfs to see if I get the same outputs as previous."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "prots = '/mnt/scgc/simon/simonsproject/jb_vs_test/AG-903/AG-903-I06/prodigal/AG-903-I06_contigs_proteins.fasta'\n",
    "contigs = '/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/AG-903/AG-903-I06_contigs.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'python recruitment_for_vs.py --threads 10 --output /mnt/scgc/simon/simonsproject/bats248_vs/diamond/ --sag-contigs /mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/AG-903/AG-903-I06_contigs.fasta /mnt/scgc/simon/simonsproject/jb_vs_test/AG-903/AG-903-I06/prodigal/AG-903-I06_contigs_proteins.fasta /mnt/scgc_nfs/ref/viral_dbs/POV.fasta.gz /mnt/scgc_nfs/ref/viral_dbs/LineP-all.fasta.gz'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"python recruitment_for_vs.py --threads 10 --output /mnt/scgc/simon/simonsproject/bats248_vs/diamond/ \\\n",
    "--sag-contigs {contigs} \\\n",
    "{prots} /mnt/scgc_nfs/ref/viral_dbs/POV.fasta.gz /mnt/scgc_nfs/ref/viral_dbs/LineP-all.fasta.gz\".format(contigs=contigs, prots=prots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AG-903-I06_contigs_proteins.dmnd\r\n",
      "AG-903-I06_contigs_proteins_mg_diamond_recruitment_tbl.csv\r\n",
      "AG-903-I06_contigs_proteins_vs_LineP-all.daa\r\n",
      "AG-903-I06_contigs_proteins_vs_LineP-all.tsv.gz\r\n",
      "AG-903-I06_contigs_proteins_vs_POV.daa\r\n",
      "AG-903-I06_contigs_proteins_vs_POV.tsv.gz\r\n",
      "pergenome\r\n"
     ]
    }
   ],
   "source": [
    "!ls /mnt/scgc/simon/simonsproject/bats248_vs/diamond/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "testtbl = pd.read_csv(\"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/AG-903-I06_contigs_proteins_mg_diamond_recruitment_tbl.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>Similarity_1.LineP.all.fr</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>Similarity_1.POV.fr</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-903-I06_NODE_1</td>\n",
       "      <td>0.001918</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>0.007610</td>\n",
       "      <td>0.007480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-903-I06_NODE_2</td>\n",
       "      <td>0.013050</td>\n",
       "      <td>0.013050</td>\n",
       "      <td>0.001685</td>\n",
       "      <td>0.001685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-903-I06_NODE_3</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>0.001012</td>\n",
       "      <td>0.001012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-903-I06_NODE_4</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.005089</td>\n",
       "      <td>0.005089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-903-I06_NODE_5</td>\n",
       "      <td>0.012137</td>\n",
       "      <td>0.012137</td>\n",
       "      <td>0.009270</td>\n",
       "      <td>0.009270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG-903-I06_NODE_6</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG-903-I06_NODE_7</td>\n",
       "      <td>0.003379</td>\n",
       "      <td>0.003387</td>\n",
       "      <td>0.024417</td>\n",
       "      <td>0.024714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG-903-I06_NODE_8</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG-903-I06_NODE_9</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.000525</td>\n",
       "      <td>0.000525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG-903-I06_NODE_10</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.002049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AG-903-I06_NODE_11</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>0.002958</td>\n",
       "      <td>0.002958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AG-903-I06_NODE_12</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.050855</td>\n",
       "      <td>0.050871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AG-903-I06_NODE_13</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AG-903-I06_NODE_14</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AG-903-I06_NODE_15</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AG-903-I06_NODE_16</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AG-903-I06_NODE_17</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AG-903-I06_NODE_18</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AG-903-I06_NODE_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AG-903-I06_NODE_20</td>\n",
       "      <td>0.002315</td>\n",
       "      <td>0.002315</td>\n",
       "      <td>0.010765</td>\n",
       "      <td>0.010765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AG-903-I06_NODE_21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig  Similarity_1.LineP.all.fr  fr_mg-bac  \\\n",
       "0    AG-903-I06_NODE_1                   0.001918   0.001915   \n",
       "1    AG-903-I06_NODE_2                   0.013050   0.013050   \n",
       "2    AG-903-I06_NODE_3                   0.018270   0.018270   \n",
       "3    AG-903-I06_NODE_4                   0.000308   0.000308   \n",
       "4    AG-903-I06_NODE_5                   0.012137   0.012137   \n",
       "5    AG-903-I06_NODE_6                   0.001548   0.001548   \n",
       "6    AG-903-I06_NODE_7                   0.003379   0.003387   \n",
       "7    AG-903-I06_NODE_8                   0.000027   0.000027   \n",
       "8    AG-903-I06_NODE_9                   0.007619   0.007619   \n",
       "9   AG-903-I06_NODE_10                   0.000208   0.000208   \n",
       "10  AG-903-I06_NODE_11                   0.004400   0.004400   \n",
       "11  AG-903-I06_NODE_12                   0.003247   0.003247   \n",
       "12  AG-903-I06_NODE_13                   0.000685   0.000685   \n",
       "13  AG-903-I06_NODE_14                   0.005915   0.005915   \n",
       "14  AG-903-I06_NODE_15                   0.000721   0.000721   \n",
       "15  AG-903-I06_NODE_16                   0.000022   0.000022   \n",
       "16  AG-903-I06_NODE_17                   0.000441   0.000441   \n",
       "17  AG-903-I06_NODE_18                   0.000070   0.000070   \n",
       "18  AG-903-I06_NODE_19                        NaN        NaN   \n",
       "19  AG-903-I06_NODE_20                   0.002315   0.002315   \n",
       "20  AG-903-I06_NODE_21                        NaN        NaN   \n",
       "\n",
       "    Similarity_1.POV.fr  fr_mg-vir  \n",
       "0              0.007610   0.007480  \n",
       "1              0.001685   0.001685  \n",
       "2              0.001012   0.001012  \n",
       "3              0.005089   0.005089  \n",
       "4              0.009270   0.009270  \n",
       "5              0.000126   0.000126  \n",
       "6              0.024417   0.024714  \n",
       "7                   NaN        NaN  \n",
       "8              0.000525   0.000525  \n",
       "9              0.002049   0.002049  \n",
       "10             0.002958   0.002958  \n",
       "11             0.050855   0.050871  \n",
       "12                  NaN        NaN  \n",
       "13             0.000069   0.000069  \n",
       "14             0.000023   0.000023  \n",
       "15                  NaN        NaN  \n",
       "16             0.000065   0.000065  \n",
       "17                  NaN        NaN  \n",
       "18                  NaN        NaN  \n",
       "19             0.010765   0.010765  \n",
       "20                  NaN        NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rvs.merge(testtbl, on='contig', how='outer')[['contig', 'Similarity_1.LineP.all.fr','fr_mg-bac','Similarity_1.POV.fr','fr_mg-vir']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Those are almost exactly the same.  Prokka ORFs generate much different recruitment results compared to prodigal generated ORFs.  Now we (I) know."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
