{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import os.path as op\n",
    "from recruitment_for_vs import *\n",
    "import glob\n",
    "import shutil\n",
    "from phage_count_table import phage_contig_table\n",
    "from recruitment_for_vs import summarize_by_contig\n",
    "from summary import virus_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "wd = \"/mnt/scgc/simon/simonsproject/bats248_vs\"\n",
    "\n",
    "to_start =  '''AG−891−A17\n",
    "AG−892−P18\n",
    "AG−893−J23\n",
    "AG−894−C07\n",
    "AG−895−P08\n",
    "AG−897−A15  \n",
    "AG−903−F19\n",
    "AG−903−I06\n",
    "AG−904−O13\n",
    "AG−907−C19\n",
    "AG−907−I10\n",
    "AG−908−F15\n",
    "AG-909-A05\n",
    "AG−910−E05\n",
    "AG−912−O18\n",
    "AG−913−C05\n",
    "AG−913−C17'''.replace(\"−\",\"-\").split()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## creating objects needed for munging mica results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# cluster map created in 180125_orf_prodigal_setup_reclustering_and_qc.ipynb\n",
    "\n",
    "out_map = \"/mnt/scgc/simon/simonsproject/bats248_vs/clustering/cluster_map.txt\"\n",
    "\n",
    "def write_cluster_map(cmap, out_map):\n",
    "    count = 0\n",
    "    with open(out_map, \"w\") as oh:\n",
    "        for k in cluster_map:\n",
    "            print(k, k, sep = \"\\t\", file=oh)\n",
    "            for val in cluster_map[k]:\n",
    "                print(val, k, sep=\"\\t\", file=oh)\n",
    "                \n",
    "def read_cluster_map(out_map):\n",
    "    cmap = {}\n",
    "    with open(out_map) as ih:\n",
    "        for l in ih:\n",
    "            vec = l.strip().split(\"\\t\")\n",
    "            cmap[vec[0]] = vec[1]\n",
    "    return cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "phage_hits = pd.concat([pd.read_csv(i) for i in glob.glob(op.join(wd, \"phage_hits*\"))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cmap = read_cluster_map(out_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sag = to_start[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "prot_fasta = op.join(wd, 'prodigal', '{}_proteins.fasta'.format(sag))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "phage_df = phage_contig_table(cmap, prot_fasta, phage_hits, outfile=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>total_orfs</th>\n",
       "      <th>viral_phage_gene_fraction</th>\n",
       "      <th>viral2_phage_gene_fraction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AG-891-A17_NODE_1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>166</td>\n",
       "      <td>0.018072</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-891-A17_NODE_10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-891-A17_NODE_11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-891-A17_NODE_12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-891-A17_NODE_13</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     p1   p2  total_orfs  viral_phage_gene_fraction  \\\n",
       "contig                                                                \n",
       "AG-891-A17_NODE_1   3.0  0.0         166                   0.018072   \n",
       "AG-891-A17_NODE_10  0.0  0.0          35                   0.000000   \n",
       "AG-891-A17_NODE_11  0.0  0.0          21                   0.000000   \n",
       "AG-891-A17_NODE_12  0.0  0.0          23                   0.000000   \n",
       "AG-891-A17_NODE_13  0.0  0.0          30                   0.000000   \n",
       "\n",
       "                    viral2_phage_gene_fraction  \n",
       "contig                                          \n",
       "AG-891-A17_NODE_1                          0.0  \n",
       "AG-891-A17_NODE_10                         0.0  \n",
       "AG-891-A17_NODE_11                         0.0  \n",
       "AG-891-A17_NODE_12                         0.0  \n",
       "AG-891-A17_NODE_13                         0.0  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phage_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "recruit_csv = op.join(wd, 'diamond','pergenome','{}_proteins_mg_diamond_recruitment_tbl.csv'.format(sag))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "recruit_df = pd.read_csv(recruit_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>hit_mg-bac</th>\n",
       "      <th>hit_mg-vir</th>\n",
       "      <th>reads_mg-vir</th>\n",
       "      <th>reads_mg-bac</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "      <th>ratio_virus_bacteria</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-891-A17_NODE_1</td>\n",
       "      <td>44265.0</td>\n",
       "      <td>4208.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>147339.0</td>\n",
       "      <td>0.036287</td>\n",
       "      <td>0.004823</td>\n",
       "      <td>0.132902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-891-A17_NODE_10</td>\n",
       "      <td>5766.0</td>\n",
       "      <td>669.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>32598.0</td>\n",
       "      <td>0.021365</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>0.162206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-891-A17_NODE_11</td>\n",
       "      <td>12122.0</td>\n",
       "      <td>993.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>24711.0</td>\n",
       "      <td>0.059251</td>\n",
       "      <td>0.006786</td>\n",
       "      <td>0.114522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-891-A17_NODE_12</td>\n",
       "      <td>8555.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>22815.0</td>\n",
       "      <td>0.045291</td>\n",
       "      <td>0.003604</td>\n",
       "      <td>0.079584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-891-A17_NODE_13</td>\n",
       "      <td>5673.0</td>\n",
       "      <td>635.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>22000.0</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>0.004874</td>\n",
       "      <td>0.156486</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               contig  hit_mg-bac  hit_mg-vir  reads_mg-vir  reads_mg-bac  \\\n",
       "0   AG-891-A17_NODE_1     44265.0      4208.0     5922080.0     8279226.0   \n",
       "1  AG-891-A17_NODE_10      5766.0       669.0     5922080.0     8279226.0   \n",
       "2  AG-891-A17_NODE_11     12122.0       993.0     5922080.0     8279226.0   \n",
       "3  AG-891-A17_NODE_12      8555.0       487.0     5922080.0     8279226.0   \n",
       "4  AG-891-A17_NODE_13      5673.0       635.0     5922080.0     8279226.0   \n",
       "\n",
       "   contig_length  fr_mg-bac  fr_mg-vir  ratio_virus_bacteria  \n",
       "0       147339.0   0.036287   0.004823              0.132902  \n",
       "1        32598.0   0.021365   0.003465              0.162206  \n",
       "2        24711.0   0.059251   0.006786              0.114522  \n",
       "3        22815.0   0.045291   0.003604              0.079584  \n",
       "4        22000.0   0.031146   0.004874              0.156486  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recruit_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "merged = recruit_df.merge(phage_df.reset_index(), on='contig', how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "result = merged.merge(virus_class(merged), on='contig', how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>hit_mg-bac</th>\n",
       "      <th>hit_mg-vir</th>\n",
       "      <th>reads_mg-vir</th>\n",
       "      <th>reads_mg-bac</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "      <th>ratio_virus_bacteria</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>total_orfs</th>\n",
       "      <th>viral_phage_gene_fraction</th>\n",
       "      <th>viral2_phage_gene_fraction</th>\n",
       "      <th>virus_class</th>\n",
       "      <th>virus_prob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-891-A17_NODE_1</td>\n",
       "      <td>44265.0</td>\n",
       "      <td>4208.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>147339.0</td>\n",
       "      <td>0.036287</td>\n",
       "      <td>0.004823</td>\n",
       "      <td>0.132902</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>166</td>\n",
       "      <td>0.018072</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-891-A17_NODE_10</td>\n",
       "      <td>5766.0</td>\n",
       "      <td>669.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>32598.0</td>\n",
       "      <td>0.021365</td>\n",
       "      <td>0.003465</td>\n",
       "      <td>0.162206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-891-A17_NODE_11</td>\n",
       "      <td>12122.0</td>\n",
       "      <td>993.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>24711.0</td>\n",
       "      <td>0.059251</td>\n",
       "      <td>0.006786</td>\n",
       "      <td>0.114522</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-891-A17_NODE_12</td>\n",
       "      <td>8555.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>22815.0</td>\n",
       "      <td>0.045291</td>\n",
       "      <td>0.003604</td>\n",
       "      <td>0.079584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-891-A17_NODE_13</td>\n",
       "      <td>5673.0</td>\n",
       "      <td>635.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>22000.0</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>0.004874</td>\n",
       "      <td>0.156486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG-891-A17_NODE_14</td>\n",
       "      <td>4447.0</td>\n",
       "      <td>384.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>20404.0</td>\n",
       "      <td>0.026325</td>\n",
       "      <td>0.003178</td>\n",
       "      <td>0.120720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG-891-A17_NODE_15</td>\n",
       "      <td>8548.0</td>\n",
       "      <td>587.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>19050.0</td>\n",
       "      <td>0.054198</td>\n",
       "      <td>0.005203</td>\n",
       "      <td>0.096004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG-891-A17_NODE_16</td>\n",
       "      <td>6930.0</td>\n",
       "      <td>430.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>17004.0</td>\n",
       "      <td>0.049226</td>\n",
       "      <td>0.004270</td>\n",
       "      <td>0.086746</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG-891-A17_NODE_17</td>\n",
       "      <td>422.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>13103.0</td>\n",
       "      <td>0.003890</td>\n",
       "      <td>0.001147</td>\n",
       "      <td>0.294844</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG-891-A17_NODE_18</td>\n",
       "      <td>3190.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>12781.0</td>\n",
       "      <td>0.030146</td>\n",
       "      <td>0.002154</td>\n",
       "      <td>0.071435</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               contig  hit_mg-bac  hit_mg-vir  reads_mg-vir  reads_mg-bac  \\\n",
       "0   AG-891-A17_NODE_1     44265.0      4208.0     5922080.0     8279226.0   \n",
       "1  AG-891-A17_NODE_10      5766.0       669.0     5922080.0     8279226.0   \n",
       "2  AG-891-A17_NODE_11     12122.0       993.0     5922080.0     8279226.0   \n",
       "3  AG-891-A17_NODE_12      8555.0       487.0     5922080.0     8279226.0   \n",
       "4  AG-891-A17_NODE_13      5673.0       635.0     5922080.0     8279226.0   \n",
       "5  AG-891-A17_NODE_14      4447.0       384.0     5922080.0     8279226.0   \n",
       "6  AG-891-A17_NODE_15      8548.0       587.0     5922080.0     8279226.0   \n",
       "7  AG-891-A17_NODE_16      6930.0       430.0     5922080.0     8279226.0   \n",
       "8  AG-891-A17_NODE_17       422.0        89.0     5922080.0     8279226.0   \n",
       "9  AG-891-A17_NODE_18      3190.0       163.0     5922080.0     8279226.0   \n",
       "\n",
       "   contig_length  fr_mg-bac  fr_mg-vir  ratio_virus_bacteria   p1   p2  \\\n",
       "0       147339.0   0.036287   0.004823              0.132902  3.0  0.0   \n",
       "1        32598.0   0.021365   0.003465              0.162206  0.0  0.0   \n",
       "2        24711.0   0.059251   0.006786              0.114522  0.0  0.0   \n",
       "3        22815.0   0.045291   0.003604              0.079584  0.0  0.0   \n",
       "4        22000.0   0.031146   0.004874              0.156486  0.0  0.0   \n",
       "5        20404.0   0.026325   0.003178              0.120720  0.0  0.0   \n",
       "6        19050.0   0.054198   0.005203              0.096004  0.0  0.0   \n",
       "7        17004.0   0.049226   0.004270              0.086746  0.0  0.0   \n",
       "8        13103.0   0.003890   0.001147              0.294844  0.0  0.0   \n",
       "9        12781.0   0.030146   0.002154              0.071435  0.0  0.0   \n",
       "\n",
       "   total_orfs  viral_phage_gene_fraction  viral2_phage_gene_fraction  \\\n",
       "0         166                   0.018072                         0.0   \n",
       "1          35                   0.000000                         0.0   \n",
       "2          21                   0.000000                         0.0   \n",
       "3          23                   0.000000                         0.0   \n",
       "4          30                   0.000000                         0.0   \n",
       "5          22                   0.000000                         0.0   \n",
       "6          14                   0.000000                         0.0   \n",
       "7          15                   0.000000                         0.0   \n",
       "8          13                   0.000000                         0.0   \n",
       "9          15                   0.000000                         0.0   \n",
       "\n",
       "   virus_class  virus_prob  \n",
       "0          0.0         1.0  \n",
       "1          0.0         1.0  \n",
       "2          0.0         1.0  \n",
       "3          0.0         1.0  \n",
       "4          0.0         1.0  \n",
       "5          0.0         1.0  \n",
       "6          0.0         1.0  \n",
       "7          0.0         1.0  \n",
       "8          0.0         1.0  \n",
       "9          0.0         1.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    os.mkdir(\"/mnt/scgc/simon/simonsproject/bats248_vs/summary/\")\n",
    "except:\n",
    "    print(\"oh well\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AG-359-B02'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allsags = [op.basename(i).split(\"_\")[0] for i in glob.glob(\"/mnt/scgc/simon/simonsproject/bats248_vs/prodigal/*_proteins.fasta\")]\n",
    "allsags[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/AG-918-A17_proteins_mg_diamond_recruitment_tbl.csv not found\n",
      "/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/AG-918-P23_proteins_mg_diamond_recruitment_tbl.csv not found\n",
      "/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/AG-894-J09_proteins_mg_diamond_recruitment_tbl.csv not found\n"
     ]
    }
   ],
   "source": [
    "not_found = []\n",
    "merge_error = []\n",
    "summaries = 0\n",
    "\n",
    "for sag in allsags:\n",
    "    prot_fasta = op.join(wd, 'prodigal', '{}_proteins.fasta'.format(sag))\n",
    "    recruit_csv = op.join(wd, 'diamond','pergenome','{}_proteins_mg_diamond_recruitment_tbl.csv'.format(sag))\n",
    "    out_csv = op.join(wd, 'summary', '{}_summary.csv'.format(sag))\n",
    "    \n",
    "    if op.exists(out_csv):\n",
    "        summaries += 1\n",
    "        continue\n",
    "    try:\n",
    "        for f in [prot_fasta, recruit_csv]:\n",
    "            try:\n",
    "                assert op.exists(f)\n",
    "            except:\n",
    "                print(\"{} not found\".format(f))\n",
    "                not_found.append(f)\n",
    "                raise OSError(\"file not found\")\n",
    "    except:\n",
    "        continue\n",
    "        \n",
    "    phage_df = phage_contig_table(cmap, prot_fasta, phage_hits, outfile=None)\n",
    "    recruit_df = pd.read_csv(recruit_csv)\n",
    "\n",
    "    merged = recruit_df.merge(phage_df.reset_index(), on='contig', how='outer')\n",
    "    try:\n",
    "        result = merged.merge(virus_class(merged), on='contig', how='outer')\n",
    "    except:\n",
    "        merge_error.append(sag)\n",
    "        continue\n",
    "    result.to_csv(out_csv, index=False)\n",
    "    summaries += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "missing_sags = [op.basename(i).split(\"_\")[0] for i in not_found]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AG-904-P07\t5\n",
      "AG-905-J06\t5\n",
      "AG-909-J16\t5\n",
      "AG-910-J09\t5\n",
      "AG-910-P20\t5\n",
      "AG-917-P16\t5\n",
      "AG-918-A17\t5\n",
      "AG-918-P23\t5\n",
      "AG-894-J09\t5\n",
      "AG-897-O20\t5\n",
      "AG-361-A09\t5\n",
      "AG-361-A19\t5\n",
      "AG-361-A20\t5\n",
      "AG-361-B11\t5\n",
      "AG-361-B18\t5\n",
      "AG-361-C03\t5\n",
      "AG-361-C05\t5\n",
      "AG-361-C13\t5\n",
      "AG-361-C16\t5\n",
      "AG-361-C19\t5\n",
      "AG-361-D10\t5\n",
      "AG-361-D11\t5\n",
      "AG-361-D21\t5\n",
      "AG-361-E07\t5\n",
      "AG-361-E13\t5\n",
      "AG-361-E16\t5\n",
      "AG-361-E22\t5\n",
      "AG-361-F03\t5\n",
      "AG-361-F05\t5\n",
      "AG-361-F07\t5\n",
      "AG-361-F19\t5\n",
      "AG-361-G08\t5\n",
      "AG-361-G09\t5\n",
      "AG-361-G16\t5\n",
      "AG-361-G19\t5\n",
      "AG-361-I03\t5\n",
      "AG-361-I04\t5\n",
      "AG-361-I08\t5\n",
      "AG-361-I22\t5\n",
      "AG-361-J02\t5\n",
      "AG-361-J07\t5\n",
      "AG-361-J09\t5\n",
      "AG-361-J10\t5\n",
      "AG-361-J14\t5\n",
      "AG-361-J17\t5\n",
      "AG-361-J22\t5\n",
      "AG-361-K16\t5\n",
      "AG-361-L02\t5\n",
      "AG-361-L07\t5\n",
      "AG-361-L13\t5\n",
      "AG-361-L15\t5\n",
      "AG-361-L16\t5\n",
      "AG-361-M04\t5\n",
      "AG-361-M09\t5\n",
      "AG-361-M10\t5\n",
      "AG-361-M13\t5\n",
      "AG-361-M14\t5\n",
      "AG-361-M19\t5\n",
      "AG-361-M20\t5\n",
      "AG-361-N03\t5\n",
      "AG-361-N04\t5\n",
      "AG-361-N11\t5\n",
      "AG-361-N17\t5\n",
      "AG-361-N21\t5\n",
      "AG-361-O03\t5\n",
      "AG-361-O10\t5\n",
      "AG-361-O16\t5\n",
      "AG-361-O17\t5\n",
      "AG-361-O18\t5\n",
      "AG-361-O19\t5\n",
      "AG-361-P10\t5\n",
      "AG-361-P15\t5\n",
      "AG-362-A02\t5\n",
      "AG-362-A03\t5\n",
      "AG-362-A08\t5\n",
      "AG-362-A16\t5\n",
      "AG-362-A20\t5\n",
      "AG-362-A21\t5\n",
      "AG-362-B02\t5\n",
      "AG-362-B05\t5\n",
      "AG-362-B06\t5\n",
      "AG-362-B07\t5\n",
      "AG-362-C08\t5\n",
      "AG-362-C11\t5\n",
      "AG-362-C13\t5\n",
      "AG-362-C14\t5\n",
      "AG-362-C15\t5\n",
      "AG-362-D02\t5\n",
      "AG-362-D05\t5\n",
      "AG-362-D07\t5\n",
      "AG-362-D09\t5\n",
      "AG-362-D18\t5\n",
      "AG-362-D20\t5\n",
      "AG-362-D22\t5\n",
      "AG-362-E13\t5\n",
      "AG-362-E16\t5\n",
      "AG-362-E18\t5\n",
      "AG-362-E22\t5\n",
      "AG-362-F02\t5\n",
      "AG-362-F07\t5\n",
      "AG-362-F11\t5\n",
      "AG-362-F13\t5\n",
      "AG-362-F23\t5\n",
      "AG-362-G06\t5\n",
      "AG-362-G09\t5\n",
      "AG-362-G11\t5\n",
      "AG-362-G13\t5\n",
      "AG-362-G16\t5\n",
      "AG-362-I02\t5\n",
      "AG-362-I05\t5\n",
      "AG-362-I10\t5\n",
      "AG-362-I11\t5\n",
      "AG-362-I21\t5\n",
      "AG-362-I22\t5\n",
      "AG-362-J06\t5\n",
      "AG-362-J07\t5\n",
      "AG-362-J13\t5\n",
      "AG-362-K05\t5\n",
      "AG-362-K10\t5\n",
      "AG-362-K11\t5\n",
      "AG-362-L07\t5\n",
      "AG-362-L17\t5\n",
      "AG-362-L22\t5\n",
      "AG-362-M14\t5\n",
      "AG-362-M15\t5\n",
      "AG-362-M19\t5\n",
      "AG-362-M20\t5\n",
      "AG-362-M23\t5\n",
      "AG-362-N08\t5\n",
      "AG-362-N14\t5\n",
      "AG-362-N15\t5\n",
      "AG-362-N21\t5\n",
      "AG-362-N23\t5\n",
      "AG-362-O09\t5\n",
      "AG-362-O19\t5\n",
      "AG-362-O21\t5\n",
      "AG-362-O22\t5\n",
      "AG-362-O23\t5\n",
      "AG-362-P02\t5\n",
      "AG-923-C10\t5\n",
      "AG-923-D17\t5\n",
      "AG-923-G18\t5\n",
      "AG-923-K10\t5\n",
      "AG-923-K23\t5\n",
      "AG-923-L09\t5\n",
      "AG-923-N21\t5\n",
      "AG-923-O07\t5\n",
      "AG-923-O15\t5\n",
      "AG-924-D03\t5\n",
      "AG-924-E23\t5\n",
      "AG-924-F16\t5\n",
      "AG-924-F18\t5\n",
      "AG-924-L10\t5\n",
      "AG-924-O08\t5\n",
      "AG-924-O09\t5\n",
      "AG-924-O23\t5\n",
      "AG-924-P10\t5\n",
      "AG-925-A03\t5\n",
      "AG-925-A11\t5\n",
      "AG-925-A18\t5\n",
      "AG-925-B05\t5\n",
      "AG-925-B10\t5\n",
      "AG-925-C07\t5\n",
      "AG-925-C22\t5\n",
      "AG-925-D06\t5\n",
      "AG-925-J08\t5\n",
      "AG-925-K07\t5\n",
      "AG-925-K19\t5\n",
      "AG-925-K23\t5\n",
      "AG-925-N07\t5\n",
      "AG-926-A22\t5\n",
      "AG-926-D18\t5\n",
      "AG-926-E07\t5\n",
      "AG-926-E21\t5\n",
      "AG-926-F07\t5\n",
      "AG-926-G19\t5\n",
      "AG-926-J13\t5\n",
      "AG-926-L20\t5\n",
      "AG-926-M16\t5\n",
      "AG-921-A02\t5\n",
      "AG-921-F11\t5\n",
      "AG-921-K16\t5\n",
      "AG-921-M10\t5\n"
     ]
    }
   ],
   "source": [
    "recruitment_required = []\n",
    "summary_generation_required = []\n",
    "\n",
    "for s in missing_sags:\n",
    "    recruit_files = glob.glob(\"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{s}*\".format(s=s))\n",
    "    print(s, len(recruit_files), sep=\"\\t\")\n",
    "    if len(recruit_files) == 0:\n",
    "        recruitment_required.append(s)\n",
    "    elif len(recruit_files) == 5:\n",
    "        summary_generation_required.append(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\" \".join(recruitment_required)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(recruitment_required)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## all SAGs where recruitment was required were for the small particle sorts; they are all now batch submitted for recruitment\n",
    "\n",
    "...recruitment has finished.  final table generation appears to have failed for all of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sag = summary_generation_required[0]\n",
    "#sag = allsags[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AG-904-P07'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bac_tsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_vs_LineP-all.tsv.gz\".format(sag)\n",
    "vir_tsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_vs_POV.tsv.gz\".format(sag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plate = \"-\".join(sag.split(\"-\")[:2])\n",
    "contig_file = \"/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/{plate}/{sag}_contigs.fasta\".format(plate=plate, sag=sag)\n",
    "assert op.exists(contig_file), \"could not find {}\".format(contig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from recruitment_for_vs import construct_recruit_tbl\n",
    "\n",
    "read_count_dict = {'vir_reads':5922080, 'bac_reads':8279226}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n"
     ]
    }
   ],
   "source": [
    "tbl = construct_recruit_tbl(vir_tsv, bac_tsv, read_count_dict, contig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>hit_mg-bac</th>\n",
       "      <th>hit_mg-vir</th>\n",
       "      <th>reads_mg-vir</th>\n",
       "      <th>reads_mg-bac</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "      <th>ratio_virus_bacteria</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-904-P07_NODE_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>4517.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              contig  hit_mg-bac  hit_mg-vir  reads_mg-vir  reads_mg-bac  \\\n",
       "0  AG-904-P07_NODE_3         NaN           1     5922080.0     8279226.0   \n",
       "\n",
       "   contig_length  fr_mg-bac  fr_mg-vir  ratio_virus_bacteria  \n",
       "0         4517.0        NaN   0.000037                   NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n",
      "doesn't look like input contig file is in gff format.  Will assume that contig name is embedded in the ORF name.\n",
      "looks like input config fiel is in fasta format.\n"
     ]
    }
   ],
   "source": [
    "read_count_dict = {'vir_reads':5922080, 'bac_reads':8279226}\n",
    "\n",
    "missing = []\n",
    "for sag in summary_generation_required:\n",
    "    bac_tsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_vs_LineP-all.tsv.gz\".format(sag)\n",
    "    vir_tsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_vs_POV.tsv.gz\".format(sag)\n",
    "    plate = \"-\".join(sag.split(\"-\")[:2])\n",
    "    contig_file = \"/mnt/scgc/simon/simonsproject/bats248_contigs/extracellular/{sag}_contigs.fasta\".format(plate=plate, sag=sag)\n",
    "    out_tbl = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{sag}__proteins_mg_diamond_recruitment_tbl.csv\".format(sag=sag)\n",
    "    if op.exists(out_tbl):\n",
    "        continue\n",
    "        \n",
    "    exists=True\n",
    "    for f in [bac_tsv, vir_tsv, contig_file]:\n",
    "\n",
    "        if not op.exists(f):\n",
    "            exists = False\n",
    "            missing.append(sag)\n",
    "            continue\n",
    "       \n",
    "    if exists == False:\n",
    "        continue\n",
    "            \n",
    "    tbl = construct_recruit_tbl(vir_tsv, bac_tsv, read_count_dict, contig_file)      \n",
    "    tbl.to_csv(out_tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "still_problematic = []\n",
    "for sag in summary_generation_required:\n",
    "    try:\n",
    "        shutil.move(\"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}__proteins_mg_diamond_recruitment_tbl.csv\".format(sag),\n",
    "               \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_mg_diamond_recruitment_tbl.csv\".format(sag))\n",
    "    except:\n",
    "        still_problematic.append(sag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AG-904-P07',\n",
       " 'AG-905-J06',\n",
       " 'AG-909-J16',\n",
       " 'AG-910-J09',\n",
       " 'AG-910-P20',\n",
       " 'AG-917-P16',\n",
       " 'AG-918-A17',\n",
       " 'AG-918-P23',\n",
       " 'AG-894-J09']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "still_problematic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/AG-918/AG-918-A17_contigs.fasta\n",
      "/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/AG-918/AG-918-P23_contigs.fasta\n",
      "/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/AG-894/AG-894-J09_contigs.fasta\n"
     ]
    }
   ],
   "source": [
    "read_count_dict = {'vir_reads':5922080, 'bac_reads':8279226}\n",
    "\n",
    "missing = []\n",
    "for sag in summary_generation_required:\n",
    "    bac_tsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_vs_LineP-all.tsv.gz\".format(sag)\n",
    "    vir_tsv = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{}_proteins_vs_POV.tsv.gz\".format(sag)\n",
    "    plate = \"-\".join(sag.split(\"-\")[:2])\n",
    "    contig_file = \"/mnt/scgc/simon/simonsproject/bats248_contigs/coassemblies/{plate}/{sag}_contigs.fasta\".format(plate=plate, sag=sag)\n",
    "    out_tbl = \"/mnt/scgc/simon/simonsproject/bats248_vs/diamond/pergenome/{sag}_proteins_mg_diamond_recruitment_tbl.csv\".format(sag=sag)\n",
    "    if op.exists(out_tbl):\n",
    "        continue\n",
    "        \n",
    "    exists=True\n",
    "    for f in [bac_tsv, vir_tsv, contig_file]:\n",
    "\n",
    "        if not op.exists(f):\n",
    "            exists = False\n",
    "            missing.append(sag)\n",
    "            print(f)\n",
    "            continue\n",
    "       \n",
    "    if exists == False:\n",
    "        continue\n",
    "            \n",
    "    tbl = construct_recruit_tbl(vir_tsv, bac_tsv, read_count_dict, contig_file)      \n",
    "    tbl.to_csv(out_tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AG-904-P07', 'AG-905-J06', 'AG-905-K20', 'AG-910-M04', 'AG-921-M10']"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sag = merge_error[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "prot_fasta = op.join(wd, 'prodigal', '{}_proteins.fasta'.format(sag))\n",
    "recruit_csv = op.join(wd, 'diamond','pergenome','{}_proteins_mg_diamond_recruitment_tbl.csv'.format(sag))\n",
    "out_csv = op.join(wd, 'summary', '{}_summary.csv'.format(sag))\n",
    "phage_df = phage_contig_table(cmap, prot_fasta, phage_hits, outfile=None)\n",
    "recruit_df = pd.read_csv(recruit_csv)\n",
    "\n",
    "merged = recruit_df.merge(phage_df.reset_index(), on='contig', how='outer')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>contig</th>\n",
       "      <th>hit_mg-bac</th>\n",
       "      <th>hit_mg-vir</th>\n",
       "      <th>reads_mg-vir</th>\n",
       "      <th>reads_mg-bac</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "      <th>ratio_virus_bacteria</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>total_orfs</th>\n",
       "      <th>viral_phage_gene_fraction</th>\n",
       "      <th>viral2_phage_gene_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>AG-904-P07_NODE_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>4517.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>AG-904-P07_NODE_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>AG-904-P07_NODE_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>AG-904-P07_NODE_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0             contig  hit_mg-bac  hit_mg-vir  reads_mg-vir  \\\n",
       "0         0.0  AG-904-P07_NODE_3         NaN         1.0     5922080.0   \n",
       "1         NaN  AG-904-P07_NODE_1         NaN         NaN           NaN   \n",
       "2         NaN  AG-904-P07_NODE_2         NaN         NaN           NaN   \n",
       "3         NaN  AG-904-P07_NODE_4         NaN         NaN           NaN   \n",
       "\n",
       "   reads_mg-bac  contig_length  fr_mg-bac  fr_mg-vir  ratio_virus_bacteria  \\\n",
       "0     8279226.0         4517.0        NaN   0.000037                   NaN   \n",
       "1           NaN            NaN        NaN        NaN                   NaN   \n",
       "2           NaN            NaN        NaN        NaN                   NaN   \n",
       "3           NaN            NaN        NaN        NaN                   NaN   \n",
       "\n",
       "    p1   p2  total_orfs  viral_phage_gene_fraction  viral2_phage_gene_fraction  \n",
       "0  0.0  0.0           3                   0.000000                         0.0  \n",
       "1  0.0  0.0           7                   0.000000                         0.0  \n",
       "2  0.0  0.0           2                   0.000000                         0.0  \n",
       "3  1.0  0.0           3                   0.333333                         0.0  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "import os.path as op\n",
    "from collections import Counter\n",
    "\n",
    "from bisect import bisect_left\n",
    "import numpy as np\n",
    "from sklearn import neighbors\n",
    "import pandas as pd\n",
    "from math import log\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def virus_class(vsdf, test_cols = ['viral_phage_gene_fraction','ratio_virus_bacteria'], \n",
    "                training_file = '/mnt/scgc_nfs/opt/viruscope/virus-training.csv', \n",
    "                k = 27, score_col='score'):\n",
    "    ''' function for determining virus class using KNeighborsClassifier\n",
    "    Args:\n",
    "        vsdf (pandas.DataFrame): viruscope dataframe having the the two calculated test columns and a contigs column included\n",
    "        test_cols (list): list of names for test columns to be used for KNeighbors Classifier\n",
    "        training_file (path): path to the training file for the knn algorithm\n",
    "        k (int): number of nearest neighbors\n",
    "        score_col(str): name of score column to use for training\n",
    "    Returns:\n",
    "        dataframe with columns: contig, virus_class, probability\n",
    "    '''\n",
    "    def _pval_by_count(clf, test_set, training_df, \n",
    "                  test_cols=['viral_phage_gene_fraction','ratio_virus_bacteria'],\n",
    "                 score_col='score'):\n",
    "        \n",
    "        ''' \n",
    "        function for calculating probability that matches the previously used R function \n",
    "        \n",
    "        Args:\n",
    "            clf: KNeighborsClassifier object\n",
    "            test_set: dataframe used to test knn algorithm on\n",
    "            training_df: dataframe used to train knn\n",
    "            test_cols: columns in above dataframes used to train the knn algorithm\n",
    "            score_col: column in training algorithm that reports the score\n",
    "        Returns: \n",
    "            list of pvalues in same order as rows in test_set dataframe\n",
    "        '''\n",
    "        \n",
    "        ns = clf.kneighbors(test[test_cols], return_distance=False)\n",
    "        pvals = []\n",
    "\n",
    "        for l in ns:\n",
    "            ndf = training_df.loc[l,:]\n",
    "            pvals.append(Counter(ndf['score']).most_common()[0][1] / len(ndf))\n",
    "        return pvals\n",
    "    \n",
    "    train_in = pd.read_table(training_file, sep = ',').dropna(axis = 0, how = 'any').reset_index()\n",
    "    train_in.rename(columns={'ratio.virus.bacteria':'ratio_virus_bacteria'}, inplace=True)\n",
    "    train_in[test_cols[0]] = np.log10(train_in[test_cols[0]] + 0.001)\n",
    "    train_in[test_cols[1]] = np.log10(train_in[test_cols[1]] + 0.001)\n",
    "    train_label = list(train_in.iloc[:,-1].values)\n",
    "    in_train = train_in[test_cols]\n",
    "    in_train.columns = test_cols\n",
    "    \n",
    "    # make the classifier and fit the training the data\n",
    "    \n",
    "    clf = neighbors.KNeighborsClassifier(k, weights='distance', algorithm='auto')\n",
    "    clf.fit(in_train, train_label)\n",
    "    \n",
    "    test = vsdf[['contig'] + test_cols].dropna()\n",
    "    if len(test) == 0:\n",
    "        return pd.DataFrame(columns=['contig','virus_class','virus_prob'])\n",
    "    test[test_cols[0]] = np.log10(test[test_cols[0]] + 0.001)\n",
    "    test[test_cols[1]] = np.log10(test[test_cols[1]] + 0.001)\n",
    "    try:\n",
    "        test['virus_class'] = clf.predict(test[test_cols])\n",
    "    except Exception as inst:\n",
    "        print(test)\n",
    "        raise inst\n",
    "    probs = clf.predict_proba(test[test_cols])\n",
    "    # built in probability function\n",
    "    test['prob 0'] = [i[0] for i in probs]\n",
    "    test['prob 1'] = [i[1] for i in probs]\n",
    "    test['virus_prob_new'] = [l['prob 0'] if l['virus_class'] == 0 else l['prob 1'] for i, l in test.iterrows()]\n",
    "    \n",
    "    test['virus_prob'] = _pval_by_count(clf, test, train_in, test_cols, score_col)\n",
    "    test['virus_prob_new'] = \n",
    "    \n",
    "    return test[['contig','virus_class','virus_prob','virus_prob_new']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>virus_class</th>\n",
       "      <th>virus_prob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig, virus_class, virus_prob]\n",
       "Index: []"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "virus_class(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "result = merged.merge(virus_class(merged), on='contig', how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>contig</th>\n",
       "      <th>hit_mg-bac</th>\n",
       "      <th>hit_mg-vir</th>\n",
       "      <th>reads_mg-vir</th>\n",
       "      <th>reads_mg-bac</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>fr_mg-bac</th>\n",
       "      <th>fr_mg-vir</th>\n",
       "      <th>ratio_virus_bacteria</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>total_orfs</th>\n",
       "      <th>viral_phage_gene_fraction</th>\n",
       "      <th>viral2_phage_gene_fraction</th>\n",
       "      <th>virus_class</th>\n",
       "      <th>virus_prob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>AG-904-P07_NODE_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5922080.0</td>\n",
       "      <td>8279226.0</td>\n",
       "      <td>4517.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>AG-904-P07_NODE_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>AG-904-P07_NODE_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>AG-904-P07_NODE_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0             contig  hit_mg-bac  hit_mg-vir  reads_mg-vir  \\\n",
       "0         0.0  AG-904-P07_NODE_3         NaN         1.0     5922080.0   \n",
       "1         NaN  AG-904-P07_NODE_1         NaN         NaN           NaN   \n",
       "2         NaN  AG-904-P07_NODE_2         NaN         NaN           NaN   \n",
       "3         NaN  AG-904-P07_NODE_4         NaN         NaN           NaN   \n",
       "\n",
       "   reads_mg-bac  contig_length  fr_mg-bac  fr_mg-vir  ratio_virus_bacteria  \\\n",
       "0     8279226.0         4517.0        NaN   0.000037                   NaN   \n",
       "1           NaN            NaN        NaN        NaN                   NaN   \n",
       "2           NaN            NaN        NaN        NaN                   NaN   \n",
       "3           NaN            NaN        NaN        NaN                   NaN   \n",
       "\n",
       "    p1   p2  total_orfs  viral_phage_gene_fraction  \\\n",
       "0  0.0  0.0           3                   0.000000   \n",
       "1  0.0  0.0           7                   0.000000   \n",
       "2  0.0  0.0           2                   0.000000   \n",
       "3  1.0  0.0           3                   0.333333   \n",
       "\n",
       "   viral2_phage_gene_fraction virus_class virus_prob  \n",
       "0                         0.0         NaN        NaN  \n",
       "1                         0.0         NaN        NaN  \n",
       "2                         0.0         NaN        NaN  \n",
       "3                         0.0         NaN        NaN  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
